### DISCUSSION TITLE
Actual behavior of `__context__` for execution modules?

### DISCUSSION THREAD

**Discussion by dbushong** [saltstack/salt#67004] [Q&A] [answered]
Per the docs "developing modules" [section about `__context__`](https://docs.saltproject.io/en/latest/topics/development/modules/developing.html#context):

> When running an execution module `__context__` persists across all module executions until the modules are refreshed; such as when [saltutil.sync_all](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.saltutil.html#salt.modules.saltutil.sync_all) or [state.apply](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.state.html#salt.modules.state.apply_) are executed.
...
> Note: Because `__context__` may or may not have been destroyed, always be sure to check for the existence of the key in `__context__` and generate the key before using it.

And indeed, when I look at some of the builtin modules, [e.g.](https://github.com/saltstack/salt/blob/66e5b97785b7de4e88cf84fafb41ff7ba3e4234e/salt/modules/chocolatey.py#L162-L167), they certainly seem to assume that it works that way (i.e. that `__context__` is useful as a cache between runs)

However, when I actually try a simple example of (re-)using `__context__`, it seems to be reset on every call, e.g.:

```
root@d245ff4b26d3:/home/salt# salt mywin mystuff.counter
mywin:
    1
root@d245ff4b26d3:/home/salt# salt mywin mystuff.counter
mywin:
    1
root@d245ff4b26d3:/home/salt# salt mywin mystuff.counter
mywin:
    1
```

And `counter.py` is:

```python
def counter():
    if "mycounter" in __context__:
        count = __context__["mycounter"]
    else:
        count = 0
    count += 1
    __context__["mycounter"] = count
    return count
```

Additionally, I turned on debug logging, and tried repeatedly running `chocolatey.chocolatey_version` that I linked above, and AFAICT from the logs, it is also never seeing any persistent values in `__context__` and is always running from scratch.

Do I have some minion config settings wrong?  (I tried with `multiprocess: False` and `True` with no difference)

Am I misunderstanding the documentation?

**Comment by max-arnold** [comment, marked as answer]
I think it might persist between multiple module calls that happen within a single run.

**Reply by dbushong** [reply]
> I think it might persist between multiple module calls that happen within a single run.

I agree that that seems to be the behavior, but it (to my reading) pretty directly contradicts the docs.  Unfortunately, my use case would rely heavily on the documented behavior.

**Reply by dwoz** [reply]
I think it would be worth trying to track down if this feature worked at some point. Possibly before the loader refactor which took place in 3004.0? If so, this should be considered a bug. If not, then we should clarify the documentation.

**Reply by mattp-** [reply]
@dwoz i think you introduced a regression with https://github.com/saltstack/salt/blame/master/salt/minion.py#L1990
gen_modules flushes all loader state, if you call that on thread return it makes it behave as @max-arnold described. looking at this now, even if it worked, it wouldn't work for multi-master. it wouldn't be too hard to fix, just attach a dict onto minionmanager as a class var, and use that. thoughts?

**Reply by mattp-** [reply]
https://github.com/saltstack/salt/pull/68055 noticed log_beacon relies on context, this is probably making it read entire file every iteration

**Reply by dwoz** [reply]
This would have broke with the loader re-factor which fixed several race conditions/threading bugs in the loader. If we want this functional again we'll need a comprehensive solution that handle multiprocessing with some locks around `__context__`.

I'm also curious what the expected behavior should be when a minion re-starts.

**Reply by dwoz** [reply]
@mattp- ^^

**Reply by mattp-** [reply]
@dwoz im not actually sure what the full expected behavior is :) i wouldnt expect something serialized to disk/it should be fresh on restart. shared state across all minion objects id probably expect to work / would make sense.
